File: server/src/main/java/com/example/schoolsite/services/impl/AuthorizationServiceImpl.java
package com.example.schoolsite.services.impl;

import com.example.schoolsite.entity.Pupil;
import com.example.schoolsite.entity.Role;
import com.example.schoolsite.entity.Teacher;
import com.example.schoolsite.entity.User;
import com.example.schoolsite.enumiration.ERole;
import com.example.schoolsite.jwt.JwtUtils;
import com.example.schoolsite.pojo.JwtResponse;
import com.example.schoolsite.pojo.LoginRequest;
import com.example.schoolsite.pojo.MessageResponse;
import com.example.schoolsite.pojo.SignUpRequest;
import com.example.schoolsite.services.AuthorizationService;
import com.example.schoolsite.workWithDatabase.repo.PupilRepository;
import com.example.schoolsite.workWithDatabase.repo.RoleRepository;
import com.example.schoolsite.workWithDatabase.repo.TeacherRepository;
import com.example.schoolsite.workWithDatabase.repo.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AuthorizationServiceImpl implements AuthorizationService {

    private final AuthenticationManager authenticationManager;

    private final UserRepository userRepository;

    private final PupilRepository pupilRepository;

    private final TeacherRepository teacherRepository;

    private final RoleRepository roleRepository;

    private final PasswordEncoder passwordEncoder;

    private final JwtUtils jwtUtils;

    @Override
    public JwtResponse authUser(LoginRequest loginRequest) {
        Authentication authentication = authenticationManager
                .authenticate(new UsernamePasswordAuthenticationToken(
                        loginRequest.getUsername(),
                        loginRequest.getPassword()));

        // формируем jwt токен
        SecurityContextHolder.getContext().setAuthentication(authentication);
        String jwt = jwtUtils.generateJwtToken(authentication);

        UserDetailsImpl userDetails = (UserDetailsImpl) authentication.getPrincipal();
        List<String> roles = userDetails.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.toList());

        if (userDetails.getId() != null) {
            return new JwtResponse(jwt,
                    userDetails.getId(),
                    userDetails.getUsername(),
                    roles);
        } else {
            return null;
        }
    }

    @Override
    public String registerUser(SignUpRequest signupRequest) {
        if (userRepository.existsByLogin(signupRequest.getLogin())) {
            return "Error: Username is exist";
        }

        Pupil pupil = pupilRepository.findByNameAndLastnameAndPatronymic(signupRequest.getName(), signupRequest.getLastname(), signupRequest.getPatronymic());
        Teacher teacher = teacherRepository.findByNameAndLastNameAndPatronymic(signupRequest.getName(), signupRequest.getLastname(), signupRequest.getPatronymic());

        if (signupRequest.getRole().equals("pupil")) {
            if (pupil == null) {
                return "Error: Pupil is not exist";
            }
        } else if (signupRequest.getRole().equals("teacher") || signupRequest.getRole().equals("Director")) {
            if (teacher == null) {
                return "Error: Teacher is not exist";
            }
        }

        User user = new User(signupRequest.getLogin(),
                passwordEncoder.encode(signupRequest.getPassword()),
                signupRequest.getStatus());

        Set<String> reqRoles = signupRequest.getRole();
        Set<Role> roles = new HashSet<>();

        if (reqRoles == null) {
            return "Error: Your role is null";
/*            Role userRole = roleRepository
                    .findByName(ERole.ROLE_USER)
                    .orElseThrow(() -> new RuntimeException("Error, Role USER is not found"));
              roles.add(userRole);*/
        } else {
            reqRoles.forEach(r -> {
                switch (r) {
                    case "pupil" -> {
                        Role adminRole = roleRepository
                                .findByName(ERole.ROLE_PUPIL)
                                .orElseThrow(() -> new RuntimeException("Error, Role Pupil is not found"));
                        roles.add(adminRole);
                    }
                    case "teacher" -> {
                        Role modRole = roleRepository
                                .findByName(ERole.ROLE_TEACHER)
                                .orElseThrow(() -> new RuntimeException("Error, Role Teacher is not found"));
                        roles.add(modRole);
                    }
                }
            });
        }
        user.setRoles(roles);
        userRepository.save(user);
        User userForId = userRepository.findByLogin(user.getLogin()).orElse(null);

        Set<Role> role = new HashSet<>();
        role.add(new Role(ERole.ROLE_PUPIL));

        if (userForId.getRoles().equals(role)) {
            pupil.setUserId(userRepository.getById(userForId.getId()));
            pupil.setEmail(signupRequest.getEmail());
            pupilRepository.save(pupil);
        } else {
            teacher.setUserId(userRepository.getById(userForId.getId()));
            teacher.setEmail(signupRequest.getEmail());
            teacherRepository.save(teacher);
        }
        return "User CREATED";
    }
}

File: server/src/main/java/com/example/schoolsite/services/impl/AuthorizationServiceImpl.kt
package com.example.schoolsite.services.impl

import com.example.schoolsite.entity.Pupil

@Service
@RequiredArgsConstructor
class AuthorizationServiceImpl : AuthorizationService {
    private val authenticationManager: AuthenticationManager? = null
    private val userRepository: UserRepository? = null
    private val pupilRepository: PupilRepository? = null
    private val teacherRepository: TeacherRepository? = null
    private val roleRepository: RoleRepository? = null
    private val passwordEncoder: PasswordEncoder? = null
    private val jwtUtils: JwtUtils? = null
    @Override
    fun authUser(loginRequest: LoginRequest): JwtResponse? {
        val authentication: Authentication = authenticationManager
                .authenticate(UsernamePasswordAuthenticationToken(
                        loginRequest.getUsername(),
                        loginRequest.getPassword()))

        // формируем jwt токен
        SecurityContextHolder.getContext().setAuthentication(authentication)
        val jwt: String = jwtUtils.generateJwtToken(authentication)
        val userDetails: UserDetailsImpl = authentication.getPrincipal()
        val roles: List<String> = userDetails.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.toList())
        return if (userDetails.getId() != null) {
            JwtResponse(jwt,
                    userDetails.getId(),
                    userDetails.getUsername(),
                    roles)
        } else {
            null
        }
    }

    @Override
    fun registerUser(signupRequest: SignUpRequest): String {
        if (userRepository.existsByLogin(signupRequest.getLogin())) {
            return "Error: Username is exist"
        }
        val pupil: Pupil = pupilRepository.findByNameAndLastnameAndPatronymic(signupRequest.getName(), signupRequest.getLastname(), signupRequest.getPatronymic())
        val teacher: Teacher = teacherRepository.findByNameAndLastNameAndPatronymic(signupRequest.getName(), signupRequest.getLastname(), signupRequest.getPatronymic())
        if (signupRequest.getRole().equals("pupil")) {
            if (pupil == null) {
                return "Error: Pupil is not exist"
            }
        } else if (signupRequest.getRole().equals("teacher") || signupRequest.getRole().equals("Director")) {
            if (teacher == null) {
                return "Error: Teacher is not exist"
            }
        }
        val user = User(signupRequest.getLogin(),
                passwordEncoder.encode(signupRequest.getPassword()),
                signupRequest.getStatus())
        val reqRoles: Set<String> = signupRequest.getRole()
        val roles: Set<Role> = HashSet()
        if (reqRoles == null) {
            return "Error: Your role is null"
            /*            Role userRole = roleRepository
                    .findByName(ERole.ROLE_USER)
                    .orElseThrow(() -> new RuntimeException("Error, Role USER is not found"));
              roles.add(userRole);*/
        } else {
            reqRoles.forEach { r ->
                when (r) {
                    "pupil" -> {
                        val adminRole: Role = roleRepository
                                .findByName(ERole.ROLE_PUPIL)
                                .orElseThrow { RuntimeException("Error, Role Pupil is not found") }
                        roles.add(adminRole)
                    }

                    "teacher" -> {
                        val modRole: Role = roleRepository
                                .findByName(ERole.ROLE_TEACHER)
                                .orElseThrow { RuntimeException("Error, Role Teacher is not found") }
                        roles.add(modRole)
                    }
                }
            }
        }
        user.setRoles(roles)
        userRepository.save(user)
        val userForId: User = userRepository.findByLogin(user.getLogin()).orElse(null)
        val role: Set<Role> = HashSet()
        role.add(Role(ERole.ROLE_PUPIL))
        if (userForId.getRoles().equals(role)) {
            pupil.setUserId(userRepository.getById(userForId.getId()))
            pupil.setEmail(signupRequest.getEmail())
            pupilRepository.save(pupil)
        } else {
            teacher.setUserId(userRepository.getById(userForId.getId()))
            teacher.setEmail(signupRequest.getEmail())
            teacherRepository.save(teacher)
        }
        return "User CREATED"
    }
}